cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/store/ApplicationMeta.java

```diff
@@ -76,4 +76,23 @@ public String toString() {
         .add("sourceControlMeta", sourceControlMeta)
         .toString();
   }
+
+  @Override
+  public boolean equals(Object o) {
+    if (this == o) {
+      return true;
+    }
+    if (!(o instanceof ApplicationMeta)) {
+      return false;
+    }
+    ApplicationMeta that = (ApplicationMeta) o;
+    return java.util.Objects.equals(id, that.id) && java.util.Objects.equals(spec, that.spec)
```

Comment:
Can `java.util.Objects` be imported in this java file?

Reply:
Import `java.util.Objects` and use `com.google.common.base.Objects` in the `public String toString()` function.

Reply:
You can use `equal()` and `hashCode()` methods from `com.google.common.base.Objects`

cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/runtime/schedule/queue/JobKey.java